gpu/test/base/gpu_executor.cu

```diff
@@ -137,11 +161,59 @@ TEST_F(GpuExecutor, CopiesDataFromGpu)
 
     cpu->copy_from(gpu.get(), 2, orig, copy);
 
-    EXPECT_EQ(5, copy[0]);
-    ASSERT_EQ(2, copy[1]);
+    EXPECT_EQ(3, copy[0]);
+    ASSERT_EQ(8, copy[1]);
     gpu->free(orig);
 }
 
+/* Properly checks if it works only when multiple GPUs exist */
+TEST_F(GpuExecutor, PreservesDeviceSettings)
+{
+    auto previous_device = gko::GpuExecutor::get_num_devices() - 1;
+    ASSERT_NO_CUDA_ERRORS(cudaSetDevice(previous_device));
+    auto orig = gpu->alloc<int>(2);
+    int current_device;
+    ASSERT_NO_CUDA_ERRORS(cudaGetDevice(&current_device));
+    ASSERT_EQ(current_device, previous_device);
+
+    gpu->free(orig);
+    ASSERT_NO_CUDA_ERRORS(cudaGetDevice(&current_device));
+    ASSERT_EQ(current_device, previous_device);
+}
+
+TEST_F(GpuExecutor, RunsOnProperDevice)
+{
+    int value = -1;
+    ASSERT_NO_CUDA_ERRORS(cudaSetDevice(0));
+    gpu2->run(ExampleOperation(value));
+    ASSERT_EQ(value, gpu2->get_device_id());
+}
+
+TEST_F(GpuExecutor, CopiesDataFromGpuToGpu)
+{
+    int copy[2];
+    auto orig = gpu->alloc<int>(2);
+    ASSERT_NO_CUDA_ERRORS(cudaSetDevice(0));
+    init_data<<<1, 1>>>(orig);
```

Comment:
I'm beginning to find these kernel launches are making the tests more difficult to read - one has to search for the definition of the kernel in the file to figure out what's happening in the test.

I suggest we instead define a helper function:

```cuda
template <typename Closure>
__global__ void run_on_gpu_kernel(Closure c) { c(); }

template <typename Closure>
void run_on_gpu(int device_id, Closure c)
{
    // set device
    run_on_gpu_kernel<<<1, 1>>>(c);
    // check failure on kernel launch
    // synchronize device and check for errors
    // restore original device
}
```

And then we can replace the above two lines with:

```cuda
run_on_gpu(0, [orig] __device__ {
    orig[0] = 3;
    orig[1] = 8;
});
```

I feel this would make the test much more readable.

Reply:
The problem is these functions are used multiple times. It's possible to reuse the functions and doing this loses that possibility. In general, if seeing the function definition of such small functions is important, there is always "/" or "*" or "Ctrl-F".

Also, you do not need all the time to set/restore devices it depends on the context, sometimes the current device was already set for some copy or w/e.

Reply:
My argument is that these shouldn't be functions in the first place (technically, they are CUDA kernels, not functions, they only look like functions). They are reused only for convenience, because you were (understandably, I would do the same) lazy to write another "init_data" kernel.

What kind of function is this, what does it do:

```c++
// set first two elements of data to magic values 5 and 2?
__global__ void init_data(int *data)
{
    data[0] = 5;
    data[1] = 2;
}
```

It's completely out of context, and only has sense when you look at it in a specific context where it's called. The only reason why it is a function is because syntax demands it.
That's one of the reasons why C++ introduced lambdas, to solve this problem of functions, which actually shouldn't be functions.

Also, unit tests are more about readability than code-reuse. You don't have to look at the definition of `run_on_gpu` to figure out what it does. You do have to look into `init_data` if you want to figure that out.

Reply:
> Also, you do not need all the time to set/restore devices it depends on the context, sometimes the current device was already set for some copy or w/e.

If that hurts the unit test somewhere, we don't have to make setting/restoring the device a part of the `run_on_gpu`. We can do that separately.

Reply:
We can also do `run_on_current_gpu`. And implement `run_on_gpu` using that.

Reply:
I'm seeing now that lambdas are only supported in 7.5 but very experimentally, a bit more in 8.0.
https://stackoverflow.com/questions/30438538/lambda-expressions-with-cuda#30441692

In addition, the gtest functions registered with `TEST_F` have some private declarations and the cuda compiler throws this kind of error:

```
error: An extended __host__ __device__ lambda cannot be defined in a member function that has private or protected access within its class ("GpuExecutor_CopiesDataFromGpuToGpu_Test")
```

So if we're going to do this, we need another approach. Maybe a functor? but that will not really bring the code inside the test.

Reply:
Now I remember that @venovako had the same issue when trying to write GPU tests for FloatX. That was a simpler problem, as there the whole test would run on GPU, so we ended up writing a new `CUDA_TEST` macro that used `TEST` internally.

However, this is not an option here, since we only need part of the test on the GPU.

The only alternative I see would be to use a custom fork of google test which defines the test body as a public function, but this is obviously not a reasonable solution.

So maybe let's drop this issue for now, at least until CUDA has better support for lambdas.